packages/eslint-mdx/src/traverse.ts

```diff
@@ -138,17 +138,20 @@ export class Traverse {
       return
     }
 
-    let children = node.children as Node[]
+    let children = (node as Parent).children as Literal[]
```

Comment:
this is probably fine, in other places I've been using

```ts
export function assertLiteral(node: Node) asserts node is Parent {
    if ("children" in node) return
    throw new TypeError("node is not a parent")
}
```

or using https://github.com/syntax-tree/unist-util-assert

for some extra runtime/type safety, but it's up to you

Reply:
What type of node will has no `value`? Maybe it is worth adding a test case.

Reply:
A fair number of parent nodes don't have `value`.
https://github.com/syntax-tree/mdast#paragraph, https://github.com/syntax-tree/mdast#blockquote, https://github.com/syntax-tree/mdast#list, https://github.com/syntax-tree/mdast#listitem, https://github.com/syntax-tree/mdast#emphasis, etc

Reply:
Oh, I just remembered that we only care about `['export', 'import', 'jsx']` https://github.com/mdx-js/eslint-mdx/blob/master/packages/eslint-mdx/src/parser.ts#L38

Personally, I think `Node` should an optional `value?: unkown`, it will make the ts code much cleaner.

Reply:
Some context, the unist types are very bareboned, they are mainly used as interfaces for specific node types to extend, unified tries to make this easy through https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions, where `type` is the discriminator which can differentiate nodes and their properties.

This is supported by https://github.com/syntax-tree/unist-util-visit-parents, https://github.com/syntax-tree/unist-util-visit, https://github.com/syntax-tree/unist-util-is, and https://github.com/syntax-tree/unist-util-assert which allow for quick and typesafe narrowing of types.

Or though handwritten predicates https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Adding an index or extra properties to `Node`, `Literal`, and/or `Parent` breaks the discriminating union, because even after they type has been narrowed, you'd still have to check if a property really exists on this node type or not, which is something we want to avoid as much as possible.

Reply:
For your particular example, something like

```ts
function assertMdxNode({type}: UnistNode): asserts node is ExportNode | ImportNode | JSXNode {
  if (type === 'export' || type === 'import' || type === 'jsx') return
  throw new TypeError("node is not an MDX node")
}
```

could be used to assert the traversal is working as expected.
Or the traversal could be refactored to use `unist-util-visit` or `unist-util-visit-parents`

Reply:
Well, it will definitely work as expected, but it just seems useless runtime checks to me because we've already confirmed that by previous codes, I decide to use custom `Node/Parent` types which extends the default ones.

Adding an optional `value?: unknown` will still work as you want, but also keep a better balance for simple usage.

Reply:
> confirmed that by previous codes

Even better!
Could this be updated to set the node type to be `MdxExportNode | MdxImportNode | MdxJsxNode` from where it has already been confirmed?

Reply:
> but also keep a better balance for simple usage

Could you expand on what you mean by this?
To me checking `type === 'some-node-type'` seems simple, but I may be missing a broader pattern you've implemented in this code base where it wouldn't work.

Reply:
For `node: Node`, I don't want runtime asserts, `'value' in node` will not guard `node` to be `Literal`, so I have to use `node as Literal`.

If we have `value?: unkown`:

```ts
if (typeof node.value === 'string') {
  // we get correct `node.value` type `string`
}
```

But now we have to use:

```ts
const assertLiteral = (node: Node): node is Literal => 'value' in node

if (assertLiteral(node) && typeof node.value === 'string') { // seems redundant to me
  // we get correct `node.value` type `string`
}
```

Reply:
> Could this be updated to set the node type to be MdxExportNode | MdxImportNode | MdxJsxNode from where it has already been confirmed?

The first question is where is `MdxExportNode | MdxImportNode | MdxJsxNode`? I have to write them manually? So verbose, I just want a `Node` with `value: string`.

Reply:
@JounQin if a `Node` is truly any kind of `unist` compatible node, `if (typeof node.value === 'string') {` isn't a safe assumption to make, it could be anything.

That said, and as you have noted, you aren't dealing with any `unist` node, you are dealing with `mdast` and `mdx` nodes, here specifically with just the `mdx` nodes.
Where/how is each `node` coming from? Could there be an opportunity to update the function/library to pass `mdast` or `mdx` node types though so that type predicates are not necessary?

---

a work around in the meantime

```ts
const assertLiteral = (node: Node): node is Literal & { value: string } => 'value' in node && typeof node.value === 'string'

if (assertLiteral(node) ) {
  // we get correct `node.value` type `string`
}
```

could be used to reduce boilerplate, until the source of where the generic `Node` is coming from can be resolved.

Reply:
```ts
const assertLiteral = (node: Node): node is Literal & { value: string } => 'value' in node && typeof node.value === 'string'
```

This will [not work](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgOQgBM8BvAXwChLRJZEU0tc4AZBVKTAGzlFSREAzgWJlKcSXABuPAK7AAXHDlIA1kggB3JJSqVsEJEPiYhQtDHacecALxwAFJpLLCJAJTKXeBCOtotgBkcKQy8kpwJlDIAOZw5PYAfHAA5LLcCqkMcD5wQSEwAJ5gwBDouWIAdBkK9nYOqdFxqdQk2NyYUHiGxvA+bmLUCBWOZhawAVzczmIecPOkEnAA9CtwWnixwPCGUN3Y8AAGPjURR3DFpXBHzUixR3pAA)

You have to cast `node` to be `Literal` in the assertion:

```ts
const assertLiteral = (node: Node): node is Literal & { value: string } => 'value' in node && typeof (node as Literal).value === 'string'
```

---

What is the downside of optional `value?: known`? It is still type safe, and `type === 'whatever'` type assertion will still work.

For codes do not care about node the specific type, optional property is very useful.

Reply:
I decide to try to add generic types into `Literal/Parent`:

```ts
export interface Literal<T = unkown> extends Node {
  value: T
}

export interface Parent<T extends Node = Node> {
  children: T[]
}
```

So that we can do:

```ts
declare parent: Parent<Literal<string>>
```

I'll raise a PR into DefinitelyTyped.

Reply:
> ```ts
> const assertLiteral = (node: Node): node is Literal & { value: string } => 'value' in node && typeof (node as Literal).value === 'string'
> ```

You're right, good catch


> What is the downside of optional value?: known? It is still type safe, and type === 'whatever' type assertion will still work.

https://github.com/unifiedjs/unified/discussions/142#discussion-3437165 is an example and has some discussion of the issues extra properties, or having an indexed type can cause
TL;DR overview, it isn't completely typesafe, it can hide build time detectable errors.

> For codes do not care about node's specific type, optional property is very useful.

I agree with you, what I've been getting at asking by "where is the node coming from?" is the potential that it may make sense for parts of unified or unist-utils to return `Node | Literal | Parent` so https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgOQgBM8BvAKDjhgE8xgAuOAZximQHNyBfc80SLEQo0WXHAAyCVFEwAbOKFRIizAsTKUqcAG7yArozj6kAayQQA7kh58B0eMhli8ABUxRgKRSGWr1JHAU2nDYABYIckSeSEyEJADaALq2fOgm2DAIEEg+mAC2YHLAABQAlHEacAA+ktJo8jVw7jHwwXCeMPpQuaS09EwA5KAFRcCDvLzk2DmsLBD5wADSyEQA8ujxeAC8eYXF5XwI6HAlAER6coZnwvOLKyobW2VBWswLy6tPGgB0l4Y8IA would be possible

Reply:
/cc @Murderlon @remcohaszing @wooorm who have also recently been working on/reviewing typings.

Reply:
> TL;DR overview, it isn't completely typesafe, it can hide build time detectable errors.

I understand `[k: string]: string` is unsafe, but `value?: unknown`? I don't think so.

> it may make sense for parts of unified or unist-utils to return Node | Literal | Parent

This is more reasonable now. But in practice, even we get `Literal` type correctly, but then what? We still only have a property `value: unkown`, nothing more, it seems still same result with `Node` with `value?: unkown`.

Reply:
> but value?: unknown? I don't think so.

It is, `Literal` (`value`) and `Parent` (`children`) are generally mutually exclusive.
```ts
parentNode.value // this should error
literalNode.children // this should also error
```

> We still only have a property `value: unknown`, nothing more, it seems still same result with Node with `value?: unknown`.

Yep, and that is why discriminated union base off `type` (https://github.com/mdx-js/eslint-mdx/pull/321#discussion_r667057384) is the preferred approach.
If you want something more specific, you probably want a specific node type, or a small set of node types.

Reply:
> parentNode.value // this should error

Right, but it is still `unknown`

> literalNode.children // this should also error

This is still an error with `value?: unkown`

---

This is what I mean `balance`. I know stricter types maybe better, but it is still maybe painful to write. Just my personal thought.

Reply:
> I know stricter types maybe better, but it is still maybe painful to write. Just my personal thought.

We don't want the types to be painful. We may want the types to be opinionated.

Trying to identify nodes based off something other than `type` is a pattern we've seen beginners run into problems/bugs with.
And with that context, I'm okay with it being a bit painful in the typings by default, I want to steer people away from that pattern.

For people who have experience/expertise with typings, such as yourself, more advanced patterns may be useful, I'm cautiously okay with these project extending the base `unist` node types to support the project's needs (as long as it doesn't modify the global types, and impact other projects).

As a middle ground what would you think of:
* adding a TypeScript guide to https://unifiedjs.com/learn introducing common patterns in unified types
* I could pull down the project locally and open PR(s) suggesting places where ideomatic unified style type narrowing may simplify the code

Would you be interested in one or both?

Reply:
I'm OK with both, thanks if you want and can help to improve the types.

Reply:
I've created a draft of a TS guide at https://github.com/unifiedjs/unifiedjs.github.io/pull/36, your thoughts and insights are welcome!
I'll pull down the latest code from `eslint-mdx` and take a look around soon.

Reply:
opened a version of what using concrete types for traversal could look like at https://github.com/mdx-js/eslint-mdx/pull/325 for further discussion